hooks/pre-push

```diff
@@ -10,25 +33,80 @@ GREEN='\033[0;32m'
 YELLOW='\033[0;33m'
 NC='\033[0m' # No Color
 
+PREFIX="${GREEN}[PREPUSH]${NC}"
+FAILURE="${RED}FAILED${NC}"
+WARNING="${RED}[WARNING]${NC}"
+SUCCESS="${GREEN}ok${NC}"
+
+printf "${PREFIX} Clean working copy ... "
 git diff --exit-code &>/dev/null
-if [ "$?" != 0 ];
+if [ "$?" == 0 ];
 then
-	echo ${RED}ERROR${NC} You have uncommitted changes please commit or stash them before pushing so that I can run tests!
-	exit 1
+	printf "${SUCCESS}\n"
+else
+	printf "${FAILURE}\n"
+	if [ "${TARPC_ALLOW_DIRTY}" == "1" ]
+	then
+		echo ${WARNING} Running with unclean working directory
+	else
+		exit 1
+	fi
 fi
 
-printf "${YELLOW}[PRESUBMIT]${NC} Running tests ... "
-
 TEST_RESULT=0
-cargo test --manifest-path tarpc/Cargo.toml &>/dev/null
-if [ "$?" != "0" ];
+
+# args:
+# 1 - cargo command to run (build/test)
+# 2 - directory name of crate to build
+# 3 - rust toolchain (nightly/stable/beta)
+function run_cargo() {
+	if [ "$1" == "build" ]
+	then
+		VERB=Building
+	else
+		VERB=Testing
+	fi
+	if [ "$3" != "" ]
+	then
+		printf "${PREFIX} $VERB $2 on $3 ... "
+		multirust run $3 cargo $1 --manifest-path $2/Cargo.toml &>/dev/null
+	else
+		printf "${PREFIX} $VERB $2 ... "
+		cargo $1 --manifest-path $2/Cargo.toml &>/dev/null
+	fi
+	if [ "$?" != "0" ];
+	then
+		printf "${FAILURE}\n"
+		#printf "${OUTPUT}\n"
+		TEST_RESULT=1
+	else
+		printf "${SUCCESS}\n"
+	fi
+}
+
+printf "${PREFIX} Checking for multirust ... "
+command -v multirust &>/dev/null
+if [ "$?" == 0 ] && [ "${TARPC_USE_CURRENT_TOOLCHAIN}" == "" ];
 then
-	printf "${RED}FAILED${NC}"
-	TEST_RESULT=1
+	printf "${SUCCESS}\n"
+
+	# No need to build on nightly, the tests will do that
```

Comment:
We do need to build on nightly, because there are dev dependencies only built when testing, so there are failure modes only when building, e.g. if an `extern crate` isn't `#[cfg(test)]`ed but it's listed under `[dev-dependencies]`